### ISSUE TITLE
API auth failing randomly

### ISSUE THREAD

**Issue by sdodson** [openshift/origin#3298]
This seems to happen relatively reliably when first provisioning a master and goes away after restarting your master an additional time. However, @thoraxe reported that the problem re-occurred after some amount of time.

Example logs from master 
```
Jun 17 20:23:54 ose3-master.example.com openshift-master[3016]: E0617 20:23:54.789748    3016 reflector.go:133] Failed to list *api.ReplicationController: the server has asked for the client to provide credentials (get replicationControllers)
Jun 17 20:23:55 ose3-master.example.com openshift-master[3016]: E0617 20:23:55.374634    3016 reflector.go:133] Failed to list *api.Build: the server has asked for the client to provide credentials (get builds)
Jun 17 20:23:55 ose3-master.example.com openshift-master[3016]: E0617 20:23:55.709449    3016 reflector.go:133] Failed to list *api.ReplicationController: the server has asked for the client to provide credentials (get replicationControllers)
```

The steps we're following to reproduce this is to provision the standard three node environment using ansible, https://github.com/detiber/openshift-ansible/ rc branch, verify that nodes show ready, then immediately create a router. As you can see no pods nor events are created.

```
[root@ose3-master ~]# oadm router --create=true   --credentials=/etc/openshift/master/openshift-router.kubeconfig --images='docker-buildvm-rhose.usersys.redhat.com:5000/openshift3/ose-${component}:${version}' --config=/etc/openshift/master/admin.kubeconfig       [56/1877]
password for stats user admin has been set to agXXvmB2BZ
deploymentconfigs/router
services/router
[root@ose3-master ~]# oc get pods
NAME      READY     REASON    RESTARTS   AGE
[root@ose3-master ~]# oc status
In project default

service kubernetes (172.30.0.2:443)

service kubernetes-ro (172.30.0.1:80)

service router (172.30.222.165:80)
  router deploys docker-buildvm-rhose.usersys.redhat.com:5000/openshift3/ose-haproxy-router:v3.0.0.0 
    #1 deployment new 3 minutes ago

To see more information about a Service or DeploymentConfig, use 'oc describe service <name>' or 'oc describe dc <name>'.
You can use 'oc get all' to see lists of each of the types described above.

[root@ose3-master ~]# oc get ev
FIRSTSEEN                         LASTSEEN                          COUNT     NAME                      KIND      SUBOBJECT   REASON      SOURCE                              MESSAGE
Wed, 17 Jun 2015 20:40:14 -0400   Wed, 17 Jun 2015 20:40:14 -0400   1         ose3-node1.example.com    Node                  starting    {kubelet ose3-node1.example.com}    Starting kubelet.
Wed, 17 Jun 2015 20:40:15 -0400   Wed, 17 Jun 2015 20:40:15 -0400   1         ose3-master.example.com   Node                  starting    {kubelet ose3-master.example.com}   Starting kubelet.
Wed, 17 Jun 2015 20:40:15 -0400   Wed, 17 Jun 2015 20:40:15 -0400   1         ose3-node1.example.com    Node                  NodeReady   {kubelet ose3-node1.example.com}    Node ose3-node1.example.com status is now: NodeReady
Wed, 17 Jun 2015 20:40:16 -0400   Wed, 17 Jun 2015 20:40:16 -0400   1         ose3-master.example.com   Node                  NodeReady   {kubelet ose3-master.example.com}   Node ose3-master.example.com status is now: NodeReady
Wed, 17 Jun 2015 20:40:17 -0400   Wed, 17 Jun 2015 20:40:17 -0400   1         ose3-node1.example.com    Node                  starting    {kubelet ose3-node1.example.com}    Starting kubelet.
Wed, 17 Jun 2015 20:40:18 -0400   Wed, 17 Jun 2015 20:40:18 -0400   1         ose3-node1.example.com    Node                  NodeReady   {kubelet ose3-node1.example.com}    Node ose3-node1.example.com status is now: NodeReady
Wed, 17 Jun 2015 20:40:18 -0400   Wed, 17 Jun 2015 20:40:18 -0400   1         ose3-master.example.com   Node                  starting    {kubelet ose3-master.example.com}   Starting kubelet.
Wed, 17 Jun 2015 20:40:19 -0400   Wed, 17 Jun 2015 20:40:19 -0400   1         ose3-master.example.com   Node                  NodeReady   {kubelet ose3-master.example.com}   Node ose3-master.example.com status is now: NodeReady
```

If we restart the master the pods are created with no further action required and the api list failure spam is gone from the master logs.

@deads2k requested that I dump secrets and serviceaccounts before and after fixing the problem. I'll post those to a gist shortly.

**Comment by danmcp** [comment]
@liggitt FYI

**Comment by sdodson** [comment]
I don't see any difference in the secrets or serviceaccounts before creating the router, after failing to create it, and then after restarting the master when it's created successfully.

http://sdodson.fedorapeople.org/issue-3298.tar.gz contains a dump of secrets, service accounts, openshift master log from the point the host was provisioned through a working router, and a copy of /etc/openshift/master.

**Comment by liggitt** [comment]
@danmcp yeah, I'm here

**Comment by liggitt** [comment]
The token obtained for the controller loop to use is gone at some later point:
```
Jun 18 00:05:38 ose3-master.example.com openshift-master[4485]: I0618 00:05:38.938055    4485 jwt.go:191] Could not retrieve token openshift-infra/replication-controller-token-rhrja for service account openshift-infra/replication-controller: secrets "replication-controller-token-rhrja" not found
Jun 18 00:05:38 ose3-master.example.com openshift-master[4485]: E0618 00:05:38.938455    4485 reflector.go:133] Failed to list *api.ReplicationController: the server has asked for the client to provide credentials (get replicationControllers)
Jun 18 00:05:39 ose3-master.example.com openshift-master[4485]: I0618 00:05:39.939210    4485 jwt.go:191] Could not retrieve token openshift-infra/replication-controller-token-rhrja for service account openshift-infra/replication-controller: secrets "replication-controller-token-rhrja" not found
Jun 18 00:05:39 ose3-master.example.com openshift-master[4485]: E0618 00:05:39.939701    4485 reflector.go:133] Failed to list *api.ReplicationController: the server has asked for the client to provide credentials (get replicationControllers)
```

**Comment by sdodson** [comment]
https://gist.github.com/sdodson/38327eb863e17160ca5a has a dump of the secrets and the log at the time

**Comment by liggitt** [comment]
Updated #3301 with additional logging in the token controller

**Comment by liggitt** [comment]
http://fpaste.org/233408/43460433/ shows secrets getting cleaned up because the token controller thinks the related service account doesn't exist any more

**Comment by deads2k** [comment]
@liggitt can you post the full log?

**Comment by deads2k** [comment]
Without actually figuring out the problem, I'd say the simplest stop-gap to ship is to add a live check here: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/serviceaccount/tokens_controller.go#L192 and in the other secret handlers that only checks during the failure cases.

**Comment by liggitt** [comment]
@sdodson confirm this is fixed by 3301?

**Comment by liggitt** [comment]
fixed by #3301

**Comment by sdodson** [comment]
@liggitt looks good so far
